### DISCUSSION TITLE
New Era: Tooling for Scala 3

### DISCUSSION THREAD

**Discussion by tgodzik** [scalameta/metals#2616] [General]
I am working on preparing a talk for the Scala 3 release party that is being organized soon and I started asking people on Twitter what are their Scala 3 tolling related questions/issues. Twitter might not be great for more complicated topics, so if anyone want to mention something we can probably discuss it better here. 

It's my hope that based on the feedback I will be able to better address specific concerns and show an overall plan for the future. If nothing major pops up I will probably just talk about the current status and our plans for the future.

**Comment by tpolecat** [comment]
Hi, in https://twitter.com/tpolecat/status/1371471262200045568 I said I would like to see

> Support for the new metaprogramming stuff. Debugging compile-time code, seeing results of inlining, etc. Also whatever is needed for staging, which I don't understand and haven't tried yet.

and you asked for more detail about what I meant. My initial thoughts were:

- At a minimum I would like to be able to see the results of inlining in Metals, maybe as a modifier+hover thing.
- I don't use a debugger very often because they have never been very good in Scala, but to the extent that debugging is supported at all it would be nice to support stepping through code that's running at compile time.
- Staged compilation (as I understand it) allows generated code to depend on runtime values, thus requiring a compilation phase at runtime. This would seem to lead to some potentially baffling debugging scenarios that need tool support at some point.

Not strictly metaprogramming, but kind of:

- I would like to see the result of match type reduction in Metals.

Thanks!

**Reply by tgodzik** [reply]
Thanks for the details! I think I have general idea about what you mean now, but I will need to dig into the details. 

> At a minimum I would like to be able to see the results of inlining in Metals, maybe as a modifier+hover thing.

With inlining we should be able soon to use JSR-45 to get more details about the inlining results, which will also be super useful for debugging.

> Staged compilation (as I understand it) allows generated code to depend on runtime values, thus requiring a compilation phase at runtime. This would seem to lead to some potentially baffling debugging scenarios that need tool support at some point.

I will need to seriously dig into that!

> I would like to see the result of match type reduction in Metals.

That is a great idea!

**Reply by smarter** [reply]
FWIW, I demo'ed support for inlining-in-the-IDE (and the many more advanced things we can do once we have that, like user-defined refactorings) in https://www.youtube.com/watch?v=1oZLZxttY8c, I hope metals will be able to replicate this stuff one day.

**Reply by tgodzik** [reply]
Thanks! That might give me some ideas.
